### PR TITLE
Add ability to ignore folders offered by other nodes

### PR DESCRIPTION
**Pull request by nrm21**
### Purpose

Add ability to ignore folders permanently (fixes request #3993)

### Screenshots

[http://i.imgur.com/LAlN5lY.png](http://i.imgur.com/LAlN5lY.png)

### Authorship

nrm21

### Notes

So here's the thing... The button still doesn't work.  I need a javascript/angular/whatever front end guy to look at it and tell me what I'm doing wrong.   It's suppose to add the folder to the config when you press the button but, it's not adding anything to the config.  When you add to the config manually (via rest API or editing the xml file in program dir) then everything works fine.  So it's just the UI code that's not working.  probably just an easy one line of code somewhere.

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
@calmh so I had this when I first added the feature. You've asked to remove this as the solution to this is to get the other side to unshare it. What has changed that we now think its a good idea?

**Comment by calmh** [comment]
Maybe just that I've become more used to asymmetric configs being a reality and have ceased to fight it. Look far enough back and it was a protocol error for the configs not to be identical.

**Comment by AudriusButkevicius** [comment]
@st-review lgtm

**Comment by st-review** [comment]
@AudriusButkevicius: Noted! Need another LGTM or explicit merge command.

**Comment by calmh** [comment]
looks good to me too but apparently doesn't work so lets have someone chime in on that first :D

File: gui/default/index.html [hunk 1]
```diff
@@ -216,6 +216,9 @@ <h3 class="panel-title">
               <button type="button" class="btn btn-sm btn-default" ng-click="dismissFolderRejection(event.data.folder, event.data.device)">
                 <span class="fa fa-clock-o"></span>&nbsp;<span translate>Later</span>
               </button>
+              <button type="button" class="btn btn-sm btn-default" ng-click="ignoreRejectedFolder(event.data.folder, event.data.device)">
+                <span class="fa fa-clock-o"></span>&nbsp;<span translate>Ignore</span>
```
Comment (uok) [inline gui/default/index.html]: use "fa-times" instead of "fa-clock-o"

**Comment by AudriusButkevicius** [comment]
well looks to me like it should work, what's did you try adding breakpoints to the javascript debugger, and are you sure you compiled the assets in the build and did a ctrl+f5 reload?

File: lib/model/model.go [hunk 2]
```diff
@@ -809,6 +809,11 @@ func (m *Model) ClusterConfig(deviceID protocol.DeviceID, cm protocol.ClusterCon
 			continue
 		}
 
+		if m.cfg.IgnoredFolder(folder.ID) {
+			l.Infof("Ignoring folder %s since we are configured to", folder.Label)
```
Comment (imsodin) [inline lib/model/model.go]: Should probably use folder.Description().

File: lib/config/config.go [hunk 3]
```diff
@@ -155,6 +155,7 @@ type Configuration struct {
 	GUI            GUIConfiguration      `xml:"gui" json:"gui"`
 	Options        OptionsConfiguration  `xml:"options" json:"options"`
 	IgnoredDevices []protocol.DeviceID   `xml:"ignoredDevice" json:"ignoredDevices"`
+	IgnoredFolders []string              `xml:"ignoredFolder" json:"ignoredFolders"`
```
Comment (calmh) [inline lib/config/config.go]: Yep, it's a matter of the structure of the generated markup.

Comment (nrm21) [inline lib/config/config.go]: I just copied the IgnoredDevices as much as I could on this one.

File: gui/default/index.html (hunk 1 above)
Comment (nrm21) [inline gui/default/index.html]: Done.

File: gui/default/syncthing/core/syncthingController.js [hunk 4]
```diff
@@ -1532,6 +1532,12 @@ angular.module('syncthing.core')
             delete $scope.folderRejections[folder + "-" + device];
         };
 
+        $scope.ignoreRejectedFolder = function (folder, device) {
+            $scope.config.ignoredFolders.push(folder.id);
```
Comment (nrm21) [inline gui/default/syncthing/core/syncthingController.js]: On why the button isn't working Could it be that I'm using folder.id here?  I do want only a string and not the whole folder struct in the portion of the config I wrote.  But maybe I'm not understanding how this feature works.

Comment (AudriusButkevicius) [inline gui/default/syncthing/core/syncthingController.js]: event.data.folder that you are passing into this function I believe is already a string, so "string".id ends up undefined.

Comment (nrm21) [inline gui/default/syncthing/core/syncthingController.js]: Ok well either that did it OR... just a rebuild did.  Don't know which but it's working now.  :D

Comment (nrm21) [inline gui/default/syncthing/core/syncthingController.js]: Ready for PR.

**Comment by calmh** [comment]
@st-review are you there?

**Comment by st-review** [comment]
I'm sorry, @calmh. I'm afraid I don't know what you mean. I know how to `don't`, `lgtm`, `merge`, `prevent`, `squash`, `stop`.

File: lib/model/model.go (hunk 2 above)
Comment (calmh) [inline lib/model/model.go]: Y, and maybe "... from device %s" as well

Comment (nrm21) [inline lib/model/model.go]: Done.

**Comment by nrm21** [comment]
Only issue with this code is... what if a user wants to remove an ignore.  I mean it's easy enough to remove from the config.xml and restart but... will users like that?

Then again I guess that's how it is for unignore of devices also.

File: lib/model/model.go (hunk 2 above)
Comment (imsodin) [inline lib/model/model.go]: s/folder.Label/folder.Description()/ ?

Comment (nrm21) [inline lib/model/model.go]: fixed.

File: lib/model/model.go [hunk 5]
```diff
@@ -809,6 +809,16 @@ func (m *Model) ClusterConfig(deviceID protocol.DeviceID, cm protocol.ClusterCon
 			continue
 		}
 
+		if m.cfg.IgnoredFolder(folder.ID) {
+			for _, dev := range folder.Devices {
+				// Find friendly name instead of dev id and report that to console
```
Comment (calmh) [inline lib/model/model.go]: Sorry, one last nitpick :/ We always log the device ID to console. (We should, at some point, gain a deviceDescription method like we have for folders, to print a short ID and label, but we don't currently have that.)

**Comment by calmh** [comment]
@st-review lgtm

gui, lib/config, lib/model: Add ability to ignore folders offered by other nodes (fixes #3993)

**Comment by st-review** [comment]
:ok_hand: Merged as b49bbe82ddf3ab0eae2c1482f0974a116ac50108. Thanks, @nrm21!

**Pull request closed without merging**